File: hackerrank-orchestrate-september26-main/code/data_loader.py
import pandas as pd
from decimal import Decimal
from typing import List, Dict, Optional
import os

from models import (
    UserProfile, FinancialEvent, ExchangeRate, RequestPaymentOption,
    MessageRecord, ImageRecord, RequestRecord, EventStatus, FlexibilityType
)

def parse_decimal(val) -> Optional[Decimal]:
    if pd.isna(val) or val is None or str(val).strip() == '':
        return None
    try:
        return Decimal(str(val).strip())
    except Exception:
        return None

def parse_list(val) -> List[str]:
    if pd.isna(val) or val is None or str(val).strip() == '':
        return []
    return [item.strip() for item in str(val).split('|') if item.strip()]

def parse_int(val) -> Optional[int]:
    if pd.isna(val) or val is None or str(val).strip() == '':
        return None
    try:
        return int(float(str(val).strip()))
    except Exception:
        return None

def parse_bool(val) -> bool:
    if pd.isna(val) or val is None:
        return False
    val_str = str(val).strip().lower()
    return val_str in ['true', '1', 't', 'yes']

class DataLoader:
    def __init__(self, dataset_dir: str):
        self.dataset_dir = dataset_dir
        self.profiles: Dict[str, UserProfile] = {}
        self.events: List[FinancialEvent] = []
        self.exchange_rates: List[ExchangeRate] = []
        self.payment_options: Dict[str, List[RequestPaymentOption]] = {}
        self.messages: List[MessageRecord] = []
        self.images: List[ImageRecord] = []

    def load_all(self):
        self.load_profiles()
        self.load_events()
        self.load_exchange_rates()
        self.load_payment_options()
        self.load_messages()
        self.load_images()

    def load_profiles(self):
        path = os.path.join(self.dataset_dir, 'financial_profiles.csv')
        df = pd.read_csv(path)
        for _, row in df.iterrows():
            prof = UserProfile(
                user_id=str(row['user_id']).strip(),
                home_currency=str(row['home_currency']).strip(),
                current_available_balance=parse_decimal(row['current_available_balance']) or Decimal('0'),
                minimum_balance_to_keep=parse_decimal(row['minimum_balance_to_keep']) or Decimal('0'),
                financial_priorities=parse_list(row.get('financial_priorities')),
                expense_categories_to_protect=parse_list(row.get('expense_categories_to_protect')),
                expense_categories_user_is_willing_to_reduce=parse_list(row.get('expense_categories_user_is_willing_to_reduce')),
                expense_categories_user_is_willing_to_stop=parse_list(row.get('expense_categories_user_is_willing_to_stop')),
                payment_methods_user_will_consider=parse_list(row.get('payment_methods_user_will_consider')),
                max_installment_months=parse_int(row.get('max_installment_months'))
            )
            self.profiles[prof.user_id] = prof

    def load_events(self):
        path = os.path.join(self.dataset_dir, 'financial_events.csv')
        df = pd.read_csv(path, dtype=str)
        for _, row in df.iterrows():
            evt = FinancialEvent(
                event_id=str(row['event_id']).strip(),
                user_id=str(row['user_id']).strip(),
                event_type=str(row['event_type']).strip(),
                description=str(row['description']).strip() if pd.notna(row['description']) else '',
                category=str(row['category']).strip() if pd.notna(row['category']) else 'other',
                direction=str(row['direction']).strip(),
                amount=parse_decimal(row['amount']),
                currency=str(row['currency']).strip(),
                event_date=str(row['event_date']).strip(),
                settlement_date=str(row['settlement_date']).strip() if pd.notna(row['settlement_date']) and str(row['settlement_date']).strip() != '' else None,
                status=EventStatus(str(row['status']).strip()),
                linked_event_id=str(row['linked_event_id']).strip() if pd.notna(row['linked_event_id']) and str(row['linked_event_id']).strip() != '' else None,
                flexibility=FlexibilityType(str(row['flexibility']).strip()) if pd.notna(row['flexibility']) and str(row['flexibility']).strip() != '' else FlexibilityType.FIXED,
                minimum_allowed_amount=parse_decimal(row['minimum_allowed_amount'])
            )
            self.events.append(evt)

    def load_exchange_rates(self):
        path = os.path.join(self.dataset_dir, 'exchange_rates.csv')
        df = pd.read_csv(path)
        for _, row in df.iterrows():
            rate = ExchangeRate(
                rate_date=str(row['rate_date']).strip(),
                from_currency=str(row['from_currency']).strip(),
                to_currency=str(row['to_currency']).strip(),
                rate=parse_decimal(row['rate']) or Decimal('1.0')
            )
            self.exchange_rates.append(rate)

    def load_payment_options(self):
        path = os.path.join(self.dataset_dir, 'request_payment_options.csv')
        df = pd.read_csv(path)
        for _, row in df.iterrows():
            opt = RequestPaymentOption(
                payment_option_id=str(row['payment_option_id']).strip(),
                request_id=str(row['request_id']).strip(),
                payment_method=str(row['payment_method']).strip(),
                payment_amount=parse_decimal(row['payment_amount']) or Decimal('0'),
                number_of_payments=parse_int(row['number_of_payments']) or 1,
                first_payment_date=str(row['first_payment_date']).strip(),
                payment_frequency_days=parse_int(row.get('payment_frequency_days')),
                financing_fee=parse_decimal(row.get('financing_fee')) or Decimal('0'),
                total_payable_amount=parse_decimal(row['total_payable_amount']) or Decimal('0')
            )
            if opt.request_id not in self.payment_options:
                self.payment_options[opt.request_id] = []
            self.payment_options[opt.request_id].append(opt)

    def load_messages(self):
        path = os.path.join(self.dataset_dir, 'messages.csv')
        df = pd.read_csv(path)
        for _, row in df.iterrows():
            msg = MessageRecord(
                message_id=str(row['message_id']).strip(),
                user_id=str(row['user_id']).strip(),
                request_id=str(row['request_id']).strip() if pd.notna(row['request_id']) and str(row['request_id']).strip() != '' else None,
                related_event_id=str(row['related_event_id']).strip() if pd.notna(row['related_event_id']) and str(row['related_event_id']).strip() != '' else None,
                sent_at=str(row['sent_at']).strip(),
                source_type=str(row['source_type']).strip(),
                message_text=str(row['message_text']).strip()
            )
            self.messages.append(msg)

    def load_images(self):
        path = os.path.join(self.dataset_dir, 'images.csv')
        df = pd.read_csv(path)
        for _, row in df.iterrows():
            img = ImageRecord(
                image_id=str(row['image_id']).strip(),
                user_id=str(row['user_id']).strip(),
                request_id=str(row['request_id']).strip() if pd.notna(row['request_id']) and str(row['request_id']).strip() != '' else None,
                related_event_id=str(row['related_event_id']).strip()
            )
            self.images.append(img)

    def load_requests(self, filename: str = 'requests.csv') -> List[RequestRecord]:
        path = os.path.join(self.dataset_dir, filename)
        df = pd.read_csv(path)
        reqs = []
        for _, row in df.iterrows():
            req = RequestRecord(
                request_id=str(row['request_id']).strip(),
                user_id=str(row['user_id']).strip(),
                request_date=str(row['request_date']).strip(),
                request_type=str(row['request_type']).strip(),
                requested_amount=parse_decimal(row['requested_amount']) or Decimal('0'),
                desired_completion_date=str(row['desired_completion_date']).strip(),
                allows_partial_payment=parse_bool(row['allows_partial_payment']),
                request_text=str(row['request_text']).strip()
            )
            reqs.append(req)
        return reqs

File: server/dataLoader.ts
import fs from 'fs';
import path from 'path';
import { parseCSV } from './csvParser.js';

export interface UserProfileData {
  user_id: string;
  home_currency: string;
  current_available_balance: number;
  minimum_balance_to_keep: number;
  financial_priorities: string[];
  expense_categories_to_protect: string[];
  expense_categories_user_is_willing_to_reduce: string[];
  expense_categories_user_is_willing_to_stop: string[];
  payment_methods_user_will_consider: string[];
  max_installment_months?: number;
}

export interface FinancialEventData {
  event_id: string;
  user_id: string;
  event_type: string;
  description: string;
  category: string;
  direction: 'debit' | 'credit';
  amount: number;
  currency: string;
  event_date: string;
  settlement_date: string;
  status: string;
  linked_event_id: string;
  flexibility: string;
  minimum_allowed_amount?: number;
}

export interface PaymentOptionData {
  request_id: string;
  payment_option_id: string;
  payment_method: string;
  first_payment_date: string;
  number_of_payments: number;
  payment_frequency_days: number;
  payment_amount: number;
  total_cost: number;
}

export interface RequestRecordData {
  request_id: string;
  user_id: string;
  request_date: string;
  request_type: string;
  requested_amount: number;
  desired_completion_date: string;
  allows_partial_payment: boolean;
  request_text: string;
  amount_safe_to_pay?: number;
  affordability_status?: string;
  recommended_payment_method?: string;
  payment_plan?: string;
  earliest_date_for_full_payment?: string;
  spending_changes_needed?: string;
  decision_explanation?: string;
}

export interface MessageData {
  message_id: string;
  user_id: string;
  message_date: string;
  sender: string;
  content: string;
}

export class DataLoader {
  datasetDir: string;
  profiles: Map<string, UserProfileData> = new Map();
  events: FinancialEventData[] = [];
  paymentOptions: Map<string, PaymentOptionData[]> = new Map();
  sampleRequests: RequestRecordData[] = [];
  requests: RequestRecordData[] = [];
  outputMap: Map<string, Record<string, string>> = new Map();
  messages: MessageData[] = [];

  constructor(datasetDir?: string) {
    this.datasetDir = datasetDir || path.resolve(process.cwd(), 'dataset');
  }

  loadAll() {
    this.loadProfiles();
    this.loadEvents();
    this.loadPaymentOptions();
    this.loadOutputs();
    this.loadSampleRequests();
    this.loadRequests();
    this.loadMessages();
  }

  private loadProfiles() {
    const filePath = path.join(this.datasetDir, 'financial_profiles.csv');
    if (!fs.existsSync(filePath)) return;
    const text = fs.readFileSync(filePath, 'utf-8');
    const rows = parseCSV(text);

    for (const r of rows) {
      if (!r.user_id) continue;
      this.profiles.set(r.user_id, {
        user_id: r.user_id,
        home_currency: r.home_currency || 'USD',
        current_available_balance: parseFloat(r.current_available_balance) || 0,
        minimum_balance_to_keep: parseFloat(r.minimum_balance_to_keep) || 0,
        financial_priorities: r.financial_priorities ? r.financial_priorities.split('|').map(s => s.trim()) : [],
        expense_categories_to_protect: r.expense_categories_to_protect ? r.expense_categories_to_protect.split('|').map(s => s.trim()) : [],
        expense_categories_user_is_willing_to_reduce: r.expense_categories_user_is_willing_to_reduce ? r.expense_categories_user_is_willing_to_reduce.split('|').map(s => s.trim()) : [],
        expense_categories_user_is_willing_to_stop: r.expense_categories_user_is_willing_to_stop ? r.expense_categories_user_is_willing_to_stop.split('|').map(s => s.trim()) : [],
        payment_methods_user_will_consider: r.payment_methods_user_will_consider ? r.payment_methods_user_will_consider.split('|').map(s => s.trim()) : ['full_payment', 'partial_payment', 'installments'],
        max_installment_months: r.max_installment_months ? parseInt(r.max_installment_months, 10) : undefined,
      });
    }
  }

  private loadEvents() {
    const filePath = path.join(this.datasetDir, 'financial_events.csv');
    if (!fs.existsSync(filePath)) return;
    const text = fs.readFileSync(filePath, 'utf-8');
    const rows = parseCSV(text);

    for (const r of rows) {
      if (!r.event_id) continue;
      this.events.push({
        event_id: r.event_id,
        user_id: r.user_id,
        event_type: r.event_type || 'expense',
        description: r.description || '',
        category: r.category || 'general',
        direction: (r.direction === 'credit' ? 'credit' : 'debit'),
        amount: parseFloat(r.amount) || 0,
        currency: r.currency || 'USD',
        event_date: r.event_date || '',
        settlement_date: r.settlement_date || r.event_date || '',
        status: r.status || 'settled',
        linked_event_id: r.linked_event_id || '',
        flexibility: r.flexibility || 'fixed',
        minimum_allowed_amount: r.minimum_allowed_amount ? parseFloat(r.minimum_allowed_amount) : undefined,
      });
    }
  }

  private loadPaymentOptions() {
    const filePath = path.join(this.datasetDir, 'request_payment_options.csv');
    if (!fs.existsSync(filePath)) return;
    const text = fs.readFileSync(filePath, 'utf-8');
    const rows = parseCSV(text);

    for (const r of rows) {
      if (!r.request_id) continue;
      const opt: PaymentOptionData = {
        request_id: r.request_id,
        payment_option_id: r.payment_option_id || '',
        payment_method: r.payment_method || 'installments',
        first_payment_date: r.first_payment_date || '',
        number_of_payments: parseInt(r.number_of_payments, 10) || 1,
        payment_frequency_days: parseInt(r.payment_frequency_days, 10) || 30,
        payment_amount: parseFloat(r.payment_amount) || 0,
        total_cost: parseFloat(r.total_cost) || 0,
      };

      if (!this.paymentOptions.has(r.request_id)) {
        this.paymentOptions.set(r.request_id, []);
      }
      this.paymentOptions.get(r.request_id)!.push(opt);
    }
  }

  private loadOutputs() {
    const filePath = path.join(this.datasetDir, 'output.csv');
    if (!fs.existsSync(filePath)) return;
    const text = fs.readFileSync(filePath, 'utf-8');
    const rows = parseCSV(text);

    for (const r of rows) {
      if (r.request_id) {
        this.outputMap.set(r.request_id, r);
      }
    }
  }

  private loadSampleRequests() {
    const filePath = path.join(this.datasetDir, 'sample_requests.csv');
    if (!fs.existsSync(filePath)) return;
    const text = fs.readFileSync(filePath, 'utf-8');
    const rows = parseCSV(text);

    for (const r of rows) {
      if (!r.request_id) continue;
      this.sampleRequests.push({
        request_id: r.request_id,
        user_id: r.user_id,
        request_date: r.request_date,
        request_type: r.request_type,
        requested_amount: parseFloat(r.requested_amount) || 0,
        desired_completion_date: r.desired_completion_date,
        allows_partial_payment: r.allows_partial_payment === 'true' || r.allows_partial_payment === '1',
        request_text: r.request_text || '',
        amount_safe_to_pay: r.amount_safe_to_pay ? parseFloat(r.amount_safe_to_pay) : 0,
        affordability_status: r.affordability_status,
        recommended_payment_method: r.recommended_payment_method,
        payment_plan: r.payment_plan,
        earliest_date_for_full_payment: r.earliest_date_for_full_payment,
        spending_changes_needed: r.spending_changes_needed,
        decision_explanation: r.decision_explanation,
      });
    }
  }

  private loadRequests() {
    const filePath = path.join(this.datasetDir, 'requests.csv');
    if (!fs.existsSync(filePath)) return;
    const text = fs.readFileSync(filePath, 'utf-8');
    const rows = parseCSV(text);

    for (const r of rows) {
      if (!r.request_id) continue;
      const out = this.outputMap.get(r.request_id);
      this.requests.push({
        request_id: r.request_id,
        user_id: r.user_id,
        request_date: r.request_date,
        request_type: r.request_type,
        requested_amount: parseFloat(r.requested_amount) || 0,
        desired_completion_date: r.desired_completion_date,
        allows_partial_payment: r.allows_partial_payment === 'true' || r.allows_partial_payment === '1',
        request_text: r.request_text || '',
        amount_safe_to_pay: out?.amount_safe_to_pay ? parseFloat(out.amount_safe_to_pay) : 0,
        affordability_status: out?.affordability_status,
        recommended_payment_method: out?.recommended_payment_method,
        payment_plan: out?.payment_plan,
        earliest_date_for_full_payment: out?.earliest_date_for_full_payment,
        spending_changes_needed: out?.spending_changes_needed,
        decision_explanation: out?.decision_explanation,
      });
    }
  }

  private loadMessages() {
    const filePath = path.join(this.datasetDir, 'messages.csv');
    if (!fs.existsSync(filePath)) return;
    const text = fs.readFileSync(filePath, 'utf-8');
    const rows = parseCSV(text);

    for (const r of rows) {
      if (!r.message_id) continue;
      this.messages.push({
        message_id: r.message_id,
        user_id: r.user_id,
        message_date: r.message_date || '',
        sender: r.sender || '',
        content: r.content || '',
      });
    }
  }
}
